File: backend/app/agent/actions.py
"""The UI action contract: what the assistant may point a customer at.

The assistant never performs a transactional action. It may only name one of these,
and the frontend decides what to render and which deterministic flow to open. Every
value here corresponds to a control that already exists on the quote page, so an
``ui_action`` is a navigation hint — never a new capability.

Deliberately absent: a ``payment`` value. Whether money is owed depends on the
company's deposit configuration and the quote's current state, both server-side facts.
Letting the model choose between "accept" and "pay" would make it guess a commercial
question from conversational context; instead it always says :data:`ACCEPT_QUOTE` and
the frontend renders "Accept & book" or "Accept & pay the deposit" from configuration
it fetched itself.

:data:`NONE` is the default when the model suggests nothing. It is not offered to the
model — a turn with no action is expressed by not calling the tool at all, which costs
nothing, rather than by spending a tool call to say "nothing".
"""

from __future__ import annotations

import enum


class UiAction(enum.StrEnum):
    """Allowlisted navigation hints. The model may select one; it can execute none."""

    NONE = "none"
    CHANGE_DATE = "change_date"
    EDIT_MOVE = "edit_move"
    ACCEPT_QUOTE = "accept_quote"
    CONTACT_COMPANY = "contact_company"


#: What the model may choose, in the order shown to it. ``NONE`` is excluded on purpose.
SELECTABLE_ACTIONS: tuple[UiAction, ...] = (
    UiAction.CHANGE_DATE,
    UiAction.EDIT_MOVE,
    UiAction.ACCEPT_QUOTE,
    UiAction.CONTACT_COMPANY,
)

#: The model-visible enum values, as JSON Schema needs them.
SELECTABLE_VALUES: tuple[str, ...] = tuple(action.value for action in SELECTABLE_ACTIONS)

File: frontend/src/lib/actions.ts
/**
 * The UI action contract, browser side.
 *
 * The assistant may name one of these; it can operate none of them. Every value maps
 * to a control that already exists on the quote page, so an action is a scroll-and-open
 * shortcut and never a separate code path — clicking the assistant's button and
 * clicking the permanent one run exactly the same handler.
 *
 * `isUiAction` is the allowlist guard: anything unrecognised renders nothing at all,
 * so a tampered or future value degrades to a plain reply rather than a broken page.
 */

import type { QuotePublic } from "./types";
import { dollars } from "./format";

/** Must match `UiAction` in app/agent/actions.py. */
export const UI_ACTIONS = [
  "none",
  "change_date",
  "edit_move",
  "accept_quote",
  "contact_company",
] as const;

export type UiAction = (typeof UI_ACTIONS)[number];

/** Which flows a quote-page control can open. `contact_company` opens none. */
export type FlowKey = "date" | "edit" | "accept";

export function isUiAction(value: unknown): value is UiAction {
  return typeof value === "string" && (UI_ACTIONS as readonly string[]).includes(value);
}

/**
 * Label for the accept control.
 *
 * The deposit comes from the quote payload, which the backend computed — the model is
 * never asked whether money is owed, and never states an amount.
 */
export function acceptLabel(quote: Pick<QuotePublic, "deposit_cents">): string {
  return quote.deposit_cents > 0
    ? `Accept & pay ${dollars(quote.deposit_cents)} deposit`
    : "Accept & book my move";
}

export interface ActionCta {
  action: UiAction;
  label: string;
  flow: FlowKey | null;
}

/**
 * The button to render for an action, or null for none.
 *
 * Returns null once the quote is no longer live: an assistant hint that arrives after
 * the customer accepted or the quote expired must not offer a control that would fail.
 */
export function ctaFor(
  action: unknown,
  quote: Pick<QuotePublic, "deposit_cents" | "status">,
): ActionCta | null {
  if (!isUiAction(action) || action === "none") return null;
  if (quote.status !== "sent") return null;

  switch (action) {
    case "change_date":
      return { action, label: "Change date", flow: "date" };
    case "edit_move":
      return { action, label: "Edit move details", flow: "edit" };
    case "accept_quote":
      return { action, label: acceptLabel(quote), flow: "accept" };
    case "contact_company":
      return { action, label: "Contact the company", flow: null };
  }
}

/** The permanent controls shown on the quote page, in order. */
export function quoteActions(
  quote: Pick<QuotePublic, "deposit_cents" | "status">,
): ActionCta[] {
  if (quote.status !== "sent") return [];
  return [
    { action: "change_date", label: "Change date", flow: "date" },
    { action: "edit_move", label: "Edit move details", flow: "edit" },
    { action: "accept_quote", label: acceptLabel(quote), flow: "accept" },
  ];
}
